### PR TITLE
Add support for service app passwords

### LINKED ISSUE #575: Ability to create service users
This is a tracking issue for when/if the following gets implemented: https://github.com/MaterializeInc/materialize/issues/20894

### PR DESCRIPTION
**Pull request by benesch**
🚨 DO NOT RELEASE UNTIL 20 JUNE 2024. 🚨 

Fix #575.

### CODE REVIEW

**Review by bobbyiliev — COMMENTED** [review]
Oh this is amazing! Thank you for putting it all together! I will be testing it out and fixing the failing tests tomorrow morning!

**Comment by benesch** [comment]
Amazing—thanks, @bobbyiliev! We should hold off on including this into a release until 20 June 2024 (when the database side changes will go out). But would _love_ your help having all the tests fixed up now, so that we're ready to go in two weeks.

**Review by bobbyiliev — COMMENTED** [review]
This looks very good! I will do some more testing and add some examples in the docs. Also tests are passing now.

**Comment by benesch** [comment]
Hell yes! Thanks so much, @bobbyiliev. If you want to do some testing against a real personal stack that supports this new feature:

```
terraform {
  required_providers {
    materialize = {
      source  = "materialize.com/devex/materialize"
      version = "0.1"
    }
  }
}

# Configuration-based authentication
provider "materialize" {
  endpoint = "https://admin.benesch.dev.cloud.materialize.com"
  cloud_endpoint = "https://api.benesch.dev.cloud.materialize.com"
  base_endpoint = "https://dev.cloud.materialize.com"
  password       = "<sign up at https://benesch.dev.console.materialize.com/>"
  default_region = "aws/us-east-1"
}

resource "materialize_app_password" "example_app_password" {
  name = "example_app_password_name"
  type = "service"
  user = "test"
  roles = ["Member"]
}
```

File: pkg/resources/resource_app_password.go [hunk 1]
```diff
@@ -27,69 +26,146 @@ func AppPassword() *schema.Resource {
...
+				Default:      "personal",
+				ValidateFunc: validation.StringInSlice([]string{"personal", "service"}, false),
+				Description:  "The type of the app password: personal or service.",
+			},
+			"user": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Description:  "The user to associate with the app password. Only valid with service-type app passwords.",
+				ValidateFunc: validation.StringDoesNotContainAny("@"),
```
[the remark this answers is not in the document]
Comment (bobbyiliev) [reply pkg/resources/resource_app_password.go]: I just extended the validate func to include those checks as well.

File: pkg/resources/resource_app_password.go [hunk 2]
```diff
@@ -102,98 +178,115 @@ func appPasswordRead(ctx context.Context, d *schema.ResourceData, meta interface
...
+		d.Set("name", token.Description)
+		d.Set("created_at", token.CreatedAt.Format(time.RFC3339))
 
-	return nil
-}
+		// We don't update secret and password because those fields can only be
+		// determined at creation time.
+	} else {
+		// TODO: only fetch the list of SSO roles once per TF provider
+		// invocation.
```
Comment (bobbyiliev) [inline pkg/resources/resource_app_password.go:205]: Good idea! Just created a tracking issue for this: https://github.com/MaterializeInc/terraform-provider-materialize/issues/579

Comment (benesch) [reply pkg/resources/resource_app_password.go:205]: Nice, thanks!

**Review by bobbyiliev — APPROVED** [review]
Done some more testing with @benesch personal stack and so far I don't see any issues! Should be good to merge on 20 June!

**Pull request merged**